lib/common/tests/agents/crm_generate_ra_key_test.c

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2022 the Pacemaker project contributors
+ *
+ * The version control history for this file may have further details.
+ *
+ * This source code is licensed under the GNU Lesser General Public License
+ * version 2.1 or later (LGPLv2.1+) WITHOUT ANY WARRANTY.
+ */
+
+#include <crm_internal.h>
+#include <crm/common/agents.h>
+
+#include <stdarg.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <setjmp.h>
+#include <cmocka.h>
+
+static void
+all_params_null(void **state) {
+    assert_null(crm_generate_ra_key(NULL, NULL, NULL));
+}
+
+static void
+some_params_null(void **state) {
+    char *retval;
+
+    retval = crm_generate_ra_key("std", "prov", NULL);
+    assert_string_equal(retval, "std:prov:");
+    free(retval);
+
+    retval = crm_generate_ra_key("std", NULL, "ty");
+    assert_string_equal(retval, "std:ty");
+    free(retval);
+
+    retval = crm_generate_ra_key(NULL, "prov", "ty");
+    assert_string_equal(retval, ":prov:ty");
+    free(retval);
+
+    retval = crm_generate_ra_key("std", NULL, NULL);
+    assert_string_equal(retval, "std:");
+    free(retval);
+
+    retval = crm_generate_ra_key(NULL, "prov", NULL);
+    assert_string_equal(retval, ":prov:");
+    free(retval);
+
+    retval = crm_generate_ra_key(NULL, NULL, "ty");
+    assert_string_equal(retval, ":ty");
+    free(retval);
```

Comment:
I would say this function is implemented incorrectly. The only valid RA specs are std:type or std:prov:type; it should return NULL for all other possibilities. Also one of the callers assumes NULL means out of memory, but it actually asserts on invalid memory, so NULL means invalid arguments (which shouldn't be possible with the current callers).

BTW would there be any way of testing a condition that asserts? It would nice to be able to verify those.

Reply:
I haven't thought of any good way to test asserts.  We could potentially mock up crm_abort, or even abort, but the problem there is that after the mocked function returns, control will flow back to the function being tested which will not be expecting that.  I think if the test were run in a different process, we could maybe do something there.  cmocka, unfortunately, goes out of its way to advertise that it doesn't use fork for anything.

Reply:
Oh, the other problem is that abort is going to be marked as noreturn, so mocking that up could have some really weird consequences.